File: engine/b/src/main/java/net/crazyinvestor/engine_b/service/TickerService.java
package net.crazyinvestor.engine_b.service;

import lombok.RequiredArgsConstructor;
import net.crazyinvestor.engine_b.dto.response.BithumbWSTickerResponse;
import net.crazyinvestor.engine_b.entity.*;
import net.crazyinvestor.engine_b.enums.ExchangeName;
import net.crazyinvestor.engine_b.repository.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.sql.Timestamp;
import java.time.LocalDateTime;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class TickerService {
    private final ExchangeRepository exchangeRepo;
    private final SymbolRepository symbolRepo;
    private final TickerRepository tickerRepo;
    private final TickerHistoryRepository tickerHistoryRepo;
    private final QuoteRepository quoteRepo;

    @Transactional
    public void saveBithumbTicker(BithumbWSTickerResponse response) {
        String exchange = ExchangeName.BITHUMB.getExchangeName();
        String symbol = response.getSymbol();
        String tickType = response.getTickType();

        int year = Integer.parseInt(response.getDate().substring(0, 4));
        int month = Integer.parseInt(response.getDate().substring(4, 6));
        int day = Integer.parseInt(response.getDate().substring(6, 8));

        int hour = Integer.parseInt(response.getTime().substring(0, 2));
        int minute = Integer.parseInt(response.getTime().substring(2, 4));

        long timestamp = Timestamp.valueOf(LocalDateTime.of(year, month, day, hour, minute)).getTime();
        long now = Timestamp.valueOf(LocalDateTime.now()).getTime();

        double openPrice = Double.parseDouble(response.getOpenPrice());
        double highPrice = Double.parseDouble(response.getHighPrice());
        double lowPrice = Double.parseDouble(response.getLowPrice());
        double closePrice = Double.parseDouble(response.getClosePrice());
        double volume = Double.parseDouble(response.getVolume());

        Exchange exchangeObj = exchangeRepo.findByName(exchange).orElseGet(() -> {
            Exchange newItem = Exchange.builder()
                        .name(exchange)
                        .build();

            exchangeRepo.save(newItem);

            return newItem;
        });

        Symbol symbolObj = symbolRepo.findByExchangeAndName(exchangeObj, symbol).orElseGet(() -> {
            Symbol newItem = Symbol.builder()
                        .exchange(exchangeObj)
                        .name(symbol)
                        .build();

            symbolRepo.save(newItem);

            return newItem;
        });

        Ticker tickerObj = tickerRepo.findBySymbolAndTickType(symbolObj, tickType).orElseGet(() -> {
            Ticker newItem = Ticker.builder()
                    .symbol(symbolObj)
                    .tickType(tickType)
                    .build();

            tickerRepo.save(newItem);

            return newItem;
        });

        // TODO: IO 성능 최악으로 예상
        // 일정 부분 JPA 캐시에서 커버해 줄 지도? 실제 성능은 측정해봐야 할 듯
        // Issue #28
        Quote quoteObj = quoteRepo.findBySymbol(symbolObj).orElseGet(() -> {
            Quote newItem = Quote.builder()
                    .symbol(symbolObj)
                    .price(closePrice)
                    .createdAt(now)
                    .updatedAt(now)
                    .build();

            quoteRepo.save(newItem);

            return newItem;
        });

        quoteObj.setPrice(closePrice);
        quoteObj.setUpdatedAt(now);

        quoteRepo.save(quoteObj);

        TickerHistory tickerHistoryObj = TickerHistory.builder()
                .ticker(tickerObj)
                .timestamp(timestamp)
                .openPrice(openPrice)
                .highPrice(highPrice)
                .lowPrice(lowPrice)
                .closePrice(closePrice)
                .volume(volume)
                .createdAt(now)
                .updatedAt(now)
                .build();

        tickerHistoryRepo.save(tickerHistoryObj);
    }
}

File: engine/a/ticker_producer_engine/src/main/kotlin/net/crazyinvestor/ticker_producer_engine/service/TickerService.kt
package net.crazyinvestor.ticker_producer_engine.service

import com.datastax.oss.driver.api.core.uuid.Uuids
import net.crazyinvestor.engine_aaa.dto.RecentCurrencyInfoDto
import net.crazyinvestor.ticker_producer_engine.dto.response.BithumbWSTickerResponse
import org.apache.kafka.clients.producer.ProducerRecord
import org.springframework.kafka.core.reactive.ReactiveKafkaProducerTemplate
import org.springframework.stereotype.Service
import reactor.util.Loggers
import java.time.LocalDateTime
import java.util.*

@Service
class TickerService(
    val reactiveKafkaProducerTemplate: ReactiveKafkaProducerTemplate<String, RecentCurrencyInfoDto>
){
    companion object {
        val logger = Loggers.getLogger(TickerService::class.java)
    }

    fun saveBithumbTicker(response: BithumbWSTickerResponse) {
        val symbol: String = response.symbol!!
        val tickType: String = response.tickType!!
        val (date, time) = response.date!! to response.time!!

        val year: Int = date.substring(0, 4).toInt()
        val month: Int = date.substring(4, 6).toInt()
        val day: Int = date.substring(6, 8).toInt()
        val hour: Int = time.substring(0, 2).toInt()
        val minute: Int = time.substring(2, 4).toInt()

        val timestamp2 = LocalDateTime.of(year, month, day, hour, minute)
        val openPrice: Double = response.openPrice!!.toDouble()
        val highPrice: Double = response.highPrice!!.toDouble()
        val lowPrice: Double = response.lowPrice!!.toDouble()
        val closePrice: Double = response.closePrice!!.toDouble()
        val volume: Double = response.volume!!.toDouble()

        val id = when(symbol){
            "BTC_KRW" -> UUID.fromString("3889ed90-5bf2-11ed-ae98-2b786da56122")
            "ETH_KRW" -> UUID.fromString("692323f0-5bf1-11ed-ae98-2b786da56122")
            else -> TODO("ticker 새로 등록하는 경우 핸들링하기. 배치 작업?")
        }!!

        val openPrice2 = openPrice.toString()
        val highPrice2 = highPrice.toString()
        val closePrice2 = closePrice.toString()
        val createdAt = LocalDateTime.now()
        val date2: Int = "${year}${month}${day}".toInt()

        val dto = RecentCurrencyInfoDto(
            id,
            symbol, // tickerName
            tickType,
            timestamp2,
            openPrice2,
            highPrice2,
            closePrice2,
            createdAt,
            date2,
            lowPrice.toString(),
            volume.toString(),
            hour
        )

        logger.debug("PRODUCE $dto")
        reactiveKafkaProducerTemplate
            .send(ProducerRecord("tick-info", dto))
            .subscribe()

        // tickerHistoryRepo.save(tickerHistoryObj)
    }


}

//val tickerHistoryObj: TickerHistory = TickerHistory(
//    1L,
//    tickerObj,
//    timestamp,
//    openPrice,
//    highPrice,
//    lowPrice,
//    closePrice,
//    volume,
//    now,
//    now
//)
//        val exchangeObj: Exchange = exchangeRepo
//            .findByName(exchange)
//            .orElseGet {
//                val newItem: Exchange = Exchange(1L, exchange)
//                exchangeRepo.save(newItem)
//                newItem
//            }
//
//        val symbolObj: Symbol = symbolRepo
//            .findByExchangeAndName(exchangeObj, symbol)
//            .orElseGet {
//                val newItem: Symbol = Symbol(
//                    null,
//                    symbol,
//                    exchangeObj,
//                    null
//                    )
//                symbolRepo.save(newItem)
//                newItem
//            }
//
//        val tickerObj: Ticker = tickerRepo
//            .findBySymbolAndTickType(symbolObj, tickType)
//            .orElseGet {
//                val newItem = Ticker(null, symbolObj, tickType)
//                tickerRepo.save(newItem)
//                newItem
//            }
//
//        val quoteObj: Quote = quoteRepo
//            .findBySymbol(symbolObj)
//            .orElseGet {
//                val newItem: Quote = Quote(
//                    1L,
//                    symbolObj,
//                    closePrice,
//                    now,
//                    now
//                )
//                quoteRepo.save(newItem)
//                newItem
//            }
//
//        quoteObj.price = closePrice
//        quoteObj.updatedAt = now
//        quoteRepo.save(quoteObj)